### DISCUSSION TITLE
Parallel compilation of the same program for multiple targets

### DISCUSSION THREAD

**Discussion by Fabrizio-Caruso** [cc65/cc65#1988] [Ideas]
In my Cross-Lib project (https://github.com/Fabrizio-Caruso/CROSS-LIB), I build the very same game code (but not the very same full code because of macros and different included library files) for up to almost 200 targets using a dozen native and cross-compilers. Almost half of which are Zilog 80 or Intel 8080 machines, that I build with Z88DK. With Z88DK I can build my code in parallel for, say a ZX Spectrum and an MSX with no risk of temporary files being overwritten (as long as the output binary has a different name).
The second largest set of targets in my project are MOS-6502, which I build with CC65 (or VBCC for a few targets). The problem I have with CC65 is speed when mass-compiling for all CC65 targets. CC65 is twice as fast as SCCZ80 compiler (part of Z88DK) but it is slower when compiling for more than 20 targets because I have not figured if or how I can build concurrently the same game.
Could this be made possible in the future?

**Comment by groessler** [comment]
Probably temp/intermediate files (-names) are clashing.
Can you provide simple example to reproduce? And a Z88DK example to show how they do it?

**Comment by Fabrizio-Caruso** [comment]
@groessler I don't know how Z88DK does it but I remember that they discussed that they implemented this.
Parallel compilation of the same source code for different targets in Z88DK just works and you have to do nothing.
You just run several compilations at the same time as you like and there is no name clash of temporary files unless you output the same names. You can do this with make -j 8 if you have a Makefile with several targets or you could just run several builds in parallel with any other means.
For example with Cross-Lib (https://github.com/Fabrizio-Caruso/CROSS-LIB) you can do:
`make -j 16 z88dk_targets -f games/bomber/Makefile.bomber`
which will work and
`make -j 16 cc65_targets -f games/bomber/Makefile.bomber`
which will fails with errors.

**Comment by Fabrizio-Caruso** [comment]
@groessler This is how it was implemented in Z88DK:
Code:
https://github.com/z88dk/z88dk/commit/845454aea420e8846022544574c7d26ae3f28db4

Discussion:
https://github.com/z88dk/z88dk/issues/1481

I am not sure I understand what they did. 
My guess is that Z88DK will just put temp files in separate directories or file names in order to avoid name clashes.

The result is fantastic for Z88DK as it has many targets.
It would be equally good for CC65 when building multi-target projects.
